### ISSUE TITLE
Segfault when importing numpy before firedrake

### ISSUE THREAD

**Issue by asb219** [firedrakeproject/firedrake#1600]
I slightly modified the [Camassa-Holm example](https://www.firedrakeproject.org/demos/camassaholm.py.html) given in the firedrake documentation:

1) I added `import numpy as np` before `from firedrake import *`
2) I added the solver parameter `'pc_factor_mat_solver_type': 'mumps'`

This is the resulting error when python tries to solve the equation:

```
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
application called MPI_Abort(MPI_COMM_WORLD, 50152059) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=50152059
:
system msg for write_line failure : Bad file descriptor
```

However, this issue is very easy to avoid: If I import numpy _after_ importing firedrake, then everything runs fine. If I do not use the `'pc_factor_mat_solver_type': 'mumps'` solver parameter everything runs fine even when numpy is imported before firedrake.

Though this bug is so easy to avoid, someone might want to look into it.

I have a Macbook Pro (2013), OS Catalina 10.15.2.

Thanks in advance,
Alex

(@colinjcotter )

**Comment by wence-** [comment]
Hmmm. Can you please run in a debugger and figure out where the crash is coming?

```
lldb -- python my-script.py
(lldb) break set -n PetscError
(lldb) r
(lldb) c
[... should run until crash occurs]
(lldb) bt full
```

**Comment by asb219** [comment]
I followed your lldb instructions but had to replace `bt full` with `bt all` (see attached file). The output is rather long.

[lldb_output.txt](https://github.com/firedrakeproject/firedrake/files/4153281/lldb_output.txt)

**Comment by wence-** [comment]
Hmm, thanks. This is a complicated one. I wonder if somehow numpy is built with a version of blas that interferes with the one MUMPS is built against (and that hence when you import numpy first, this shadows the MUMPS library).

Can you show the output of 

```
otool -L path/to/firedrake/venv/lib/pythonXX/site-packages/numpy/linalg/lapack_lite.*.so
```

And also
```
otool -L path/to/firedrake/venv/src/petsc/default/lib/libpetsc.dylib
```

Thanks

**Comment by wence-** [comment]
I should note in the first, the pythonXX should be replaced by whatever python version your venv has (e.g. for me it is python3.7)

**Comment by asb219** [comment]
Output for the first line:

```
lib/python3.7/site-packages/numpy/linalg/lapack_lite.cpython-37m-darwin.so:
	@loader_path/../.dylibs/libopenblasp-r0.3.7.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

Output for the second line:

```
src/petsc/default/lib/libpetsc.dylib:
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libpetsc.3.012.dylib (compatibility version 3.12.0, current version 3.12.0)
	@rpath/libHYPRE-2.18.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libsuperlu_dist.6.dylib (compatibility version 6.0.0, current version 6.1.1)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libexodus.2.dylib (compatibility version 2.0.0, current version 2.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libnetcdf.13.dylib (compatibility version 14.0.0, current version 14.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libpnetcdf.3.dylib (compatibility version 4.0.0, current version 4.2.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libhdf5hl_fortran.100.dylib (compatibility version 101.0.0, current version 101.4.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libhdf5_fortran.102.dylib (compatibility version 103.0.0, current version 103.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libhdf5_hl.100.dylib (compatibility version 102.0.0, current version 102.2.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libhdf5.103.dylib (compatibility version 105.0.0, current version 105.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libpragmatic.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/X11/lib/libX11.6.dylib (compatibility version 10.0.0, current version 10.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 800.7.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libmpifort.12.dylib (compatibility version 14.0.0, current version 14.7.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libmpi.12.dylib (compatibility version 14.0.0, current version 14.7.0)
	/Users/alexanderbrunmayr/firedrake/src/petsc/default/lib/libpmpi.12.dylib (compatibility version 14.0.0, current version 14.7.0)
	/usr/local/opt/gcc/lib/gcc/9/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/usr/local/opt/gcc/lib/gcc/9/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
```

**Comment by wence-** [comment]
Hmm. I think I need someone with Catalina to try and reproduce. I can't do so on the previous version. 

What I see though is that petsc and mumps link against apple's veclib version of lapack. Whereas numpy uses openblas. Plausibly there are abi incompatibilities there

**Comment by sv2518** [comment]
Probably related to #1784

**Comment by wence-** [comment]
Fixed in #1885.
